// ion-js/dist/commonjs/es6/IonCatalog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IonSystemSymbolTable_1 = require("./IonSystemSymbolTable");
const IonUtilities_1 = require("./IonUtilities");
const IonUtilities_2 = require("./IonUtilities");
function byVersion(x, y) {
    return x.version - y.version;
}
class Catalog {
    constructor() {
        this.symbolTables = {};
        this.addSymbolTable(IonSystemSymbolTable_1.getSystemSymbolTable());
    }
    addSymbolTable(symbolTable) {
        let versions = this.symbolTables[symbolTable.name];
        if (IonUtilities_1.isUndefined(versions)) {
            versions = [];
            this.symbolTables[symbolTable.name] = versions;
        }
        versions[symbolTable.version] = symbolTable;
    }
    findSpecificVersion(name, version) {
        let versions = this.symbolTables[name];
        return (versions && versions[version]) || undefined;
    }
    findLatestVersion(name) {
        return IonUtilities_2.max(this.symbolTables[name], byVersion);
    }
}
exports.Catalog = Catalog;
